src/librustc_typeck/check/wfcheck.rs

```diff
@@ -383,8 +383,18 @@ impl<'a, 'gcx> CheckTypeWellFormedVisitor<'a, 'gcx> {
                                                   fcx.body_id,
                                                   &trait_ref,
                                                   ast_trait_ref.path.span);
+
+                    // not registering predicates associcated with a `default impl`
```

Comment:
OK, so, I think what this is doing is trying to skip the logic that requires a trait's where-clauses to be satisfied for it to be implemented. Which seems .. maybe right, maybe not. But what I don't know is why it is only skipping `trait` obligations, that seems definitely odd.

For example:

```rust
trait Foo<'a, T: Eq + 'a> { }

default impl<U> Foo<'static, U> for () { }
```

This will error because `U: 'static` is not known to hold, but will not error about the fact that `U: Eq` is not known to hold. Neither is obviously right to me. I suspect we should just not skip *any* checks here. Do you know what motivated you to do so?

Reply:
Yes, I skip registering `default impl` obligations as a conseguence of not emitting `default impl`s inside [for_each_relevat_impl](https://github.com/giannicic/rust/blob/d6cf35161444e2bb09b115d2670f485c74895d19/src/librustc/traits/select.rs#L1695.)
At first I've just skipped all kind of obligations but I noticed that, when running the full test suite, some tests failed. 
Instead when skipping only trait obligations the tests run correctly.

If I remove the `default` keyword in your example the error given is 
```
the trait bound `U: std::cmp::Eq` is not satisfied`
```
If I understand you correctly this should be the error that I should get if i put the `default` keyword again so, I'm finding out which is the obligation involved in this check in order
to extract a general rule to follow when registering these predicates.

Perhaps I should skip only the `TraitPredicate` corresponding to the `TraitRef` of the impl. I'm doing some checks and let you know

Thanks

Reply:
> At first I've just skipped all kind of obligations but I noticed that, when running the full test suite, some tests failed.

Do you mean you just *always* skipped all kinds of obligations? Or only when you skipped all obligations for `default` impls? If the latter, can you tell me which tests failed?

Still, yes, I think we should be skipping *no* obligations here, and checking the same conditions on `default` impls as non-default impls. 

>  I'm finding out which is the obligation involved in this check in order
to extract a general rule to follow when registering these predicates.

I don't understand this sentence, sorry :(